CORE-10606 schema_registry: get/post/delete_security_acls

pgellert opened the pull request:
Implement the CRUD APIs used for creating/listing/deleting ACLs associated
with schema registry resources.

Fixes https://redpandadata.atlassian.net/browse/CORE-10606

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
schema_registry/api: update swagger for acls api

* fix typo in PREFIXED resouce type
* return 200 and list of acls on delete
* change API to remove principal_type field

pgellert committed:
schema_registry: get/post/delete_security_acls

Implement the APIs used for creating/listing/deleting ACLs associated
with schema registry resources.

pgellert committed:
dt/schema_registry: add SchemaRegistryACLTest

Test the schema registry ACL CRUD endpoints through ducktape.

pgellert committed:
schema_registry: acl error codes

Introduce the error codes to be used by the ACL CRUD endpoints

pgellert committed:
schema_registry: acl request json ser/de

Define the acl types used for parsing and serializing ACL CRUD requests
and responses. Define the json serde methods on it and add a conversion
method from acl_bindings.

The `acl` holder state is reused as the representation of GET response
values, POST request values and DELETE request values as well as DELETE
response values to reduce duplication.

pgellert committed:
security/acls: add to/from_string_view

Add helpers for conversion to/from strings

pgellert committed:
schema_registry: feature gate schema_registry_authz

Introduces the feature schema_registry_authz to gate mutating ACL APIs.

This is to avoid using the new versions of the ACL-related serde types
on RPC calls and on disk until all the brokers are upgraded to the new
major version.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/4b516779dc6a39f557a8dbc502b53d4830a9014e..7784a0779988b0f4d8b8e5167725879e52192606): rebase to dev now that https://github.com/redpanda-data/redpanda/pull/26466 is merged

michael-redpanda commented on src/v/security/acl.cc:
```diff
@@ -286,82 +286,174 @@ acl_store::reset_bindings(const fragmented_vector<acl_binding>& bindings) {
 }
 
-std::ostream& operator<<(std::ostream& os, acl_operation op) {
+std::string_view to_string_view(resource_type type) {
```
nit: I think normally we make these `to_string_view` functions `constexpr`

michael-redpanda commented on src/v/pandaproxy/schema_registry/requests/acls.h:
```diff
@@ -0,0 +1,286 @@
+        host,
+        operation,
+        permission,
+    };
```
Do we want to do any validation of input when parsing out the JSON object?

e.g. should `principal_type` here be an enum of valid entires (user & redpanda_role) and should parsing fail if the provided user input does not validate against the enum?

michael-redpanda replied:
Additionally if we could split out the parsing from the handlers it would make it easier to unit test and check coverage

pgellert replied:
I've moved most of the JSON schema-style validation logic over to the JSON parsing side. I've left the tests as is in ducktape though.

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +842,189 @@ status_ready(server::request_t rq, server::reply_t rp) {
+        throw ss::httpd::bad_param_exception(
+          "Both 'principal' and 'principal_type' must be specified together, "
+          "or neither.");
+    }
```
Not trying to re-litigate the design here but do we want to just have the user specify the type as part of the principal.  This is similar to how the Kafka protocol does it.  E.g. a user principal is `User:Principal` and a role is `RedpandaRole:RoleName`.

`rpk` can insert the type automagically (like how it does kafka acls).

michael-redpanda replied:
And we can default to assuming it's a user if `RedpandaRole:` or `User:` isn't prefixing the principal name (again similar to how we handle things on the kafka side).

pgellert replied:
Yeah, I think this makes sense, since this way we can simplify the logic and share most of the rpk-client-side and redpanda-server-side logic. I've updated this now.

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +842,189 @@ status_ready(server::request_t rq, server::reply_t rp) {
+          security::resource_pattern{resource_type, acl.resource, pattern_type},
+          security::acl_entry{
+            security::acl_principal{
+              to_principal_type(acl.principal_type), acl.principal},
```
FYI I don't think we permit wildcard roles (`RedpandaRole:*`).  See https://github.com/redpanda-data/redpanda/blob/8cb6455e866164f31fd8a7a715831952334366b1/src/v/kafka/server/handlers/details/security.h#L52-L59

We don't apparently check that in the `security::acl_principal` constructor so I wonder if that's a miss there.

michael-redpanda replied:
I think this lends a little bit of credence to encoding the principal type in with the principal.

pgellert replied:
Done, the validation logic is now shared.

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +842,189 @@ status_ready(server::request_t rq, server::reply_t rp) {
+
+    auto raw_acls = co_await rjson_parse(*rq.req, acl_handler<>{});
+
+    if (raw_acls.empty()) {
+        throw exception(
+          error_code::acl_invalid, "No ACLs were specified in the request");
+    }
```
question: Should this be an error or should we just return an empty response?

pgellert replied:
I think either would be fine. I went with checking for empty explicitly for UX, since creating/deleting an empty list of ACLs seems like a caller-side error that is worth highlighting to the user. But I'm open to changing this to simplify.

michael-redpanda replied:
I think we return success on the kafka side when the list is empty - i think it's fine to return success here...

pgellert replied:
Done, I've removed the emptiness check.

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -945,6 +957,8 @@ post_security_acls(server::request_t rq, server::reply_t rp) {
             to_acl_permission(acl.permission)});
     }
 
+    check_feature_ready(rq);
```
do we need this in `get` as well?  I'd think no because it's not doing any serde stuff, right?

pgellert replied:
Exactly, it is just reading shard-local state, so we don't need it there.

michael-redpanda commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -918,6 +918,12 @@ def __init__(self,
         http.client.HTTPConnection.debuglevel = 1
         http.client.print = lambda *args: self.logger.debug(" ".join(args))
 
+    def assert_equal(self, first, second, msg=None):
+        assert first == second, msg or f"{first} != {second}"
+
+    def assert_in(self, member, container, msg=None):
+        assert member in container, msg or f"{member!r} not found in {container!r}"
```
niiiiiiice

michael-redpanda commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -5661,3 +5667,456 @@ def test_compatible_schemas(self, schema_type: str, mode: str):
+        self.assert_equal(len(resp.json()), 1)
+
+    @cluster(num_nodes=3)
+    def test_acl_filters(self):
```
nit: name, maybe `test_acl_delete_filters`

michael-redpanda reviewed (commented):
looks nice so far!

pgellert committed:
security/acls: reuse principal formatting/parsing

Reuse the principal type + name parsing and formatting logic across
Kafka and Schema Registry by moving it into the security module.

The formatting is implemented using a custom fmt formatter that supports
two representation types. This allows us to support both the Kafka/SR
API-specific formatting and the audit logging format at the same time.

I chose not to change the audit logging format to avoid introducing a
potentially breaking change, in case customers are already parsing
acl_principal. This could be reconsidered in the future to unify the
formatting.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/7784a0779988b0f4d8b8e5167725879e52192606..5932989f3966ff2a4e33351b69c4233956daef26): address code review feedback

michael-redpanda commented on src/v/security/acl.h:
```diff
@@ -291,6 +305,44 @@ class acl_principal final
+        case principal_type::role:
+            return fmt::format("RedpandaRole:{}", _name);
+        }
+        __builtin_unreachable();
```
nit: There's already an `ostream operator<<` for principal type so I think you should use that.

michael-redpanda replied:
Also why not either define an operator<< for this or use a `fmt::formatter`?

pgellert replied:
I went with changing the existing `operator<<`'s since they have only been used for logging, therefore they are safe to change.

michael-redpanda commented on src/v/pandaproxy/schema_registry/error.cc:
```diff
@@ -84,6 +84,10 @@ struct error_category final : std::error_category {
         case error_code::format_not_supported:
             return "Format parameter not supported";
+        case error_code::acl_invalid:
+            return "Invalid ACL.";
```
nit: ends with period

michael-redpanda commented on src/v/pandaproxy/schema_registry/BUILD:
```diff
@@ -152,6 +152,7 @@ redpanda_cc_library(
         "configuration.h",
         "fwd.h",
         "handlers.h",
+        "requests/acls.h",
```
nit - should this be in the previous commit?

michael-redpanda reviewed (commented):
need to reivew the tests still

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--59c49910-e5be-47d7-bd71-4fc46850774f-->
<!--TEST RESULTS-->
<!--e2406f2d-7324-483a-a9e1-85aa99e1322b-->
<!--TEST RESULTS-->
<!--23f71bbc-b5d1-403b-96ef-f797597f5dfc-->
<!--TEST RESULTS-->
<!--3dd8566e-0bcf-4db4-91b0-7492a8d55bdd-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67572">build#67572</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------|:--------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| IcebergUsageTest           | test_iceberg_usage                                | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"}                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67572#01978429-e59a-47c8-92ee-1e46a31c2193 | FLAKY         | 19/21    | upstream reliability is '84.5'. current run reliability is '90.47619047619048'. drift is -5.97619 and the allowed drift is set to 50. The test should PASS              |
| RestCatalogConnectionTest  | test_redpanda_connection_to_rest_catalog          | {"cloud_storage_type": 1}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67572#0197842c-1b3e-4e75-b3ec-e64aa9388afb | FLAKY         | 19/21    | upstream reliability is '96.46464646464646'. current run reliability is '90.47619047619048'. drift is 5.98846 and the allowed drift is set to 50. The test should PASS  |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67572#01978429-e598-4244-94b0-ec5caaa77113 | FLAKY         | 17/21    | upstream reliability is '92.61744966442953'. current run reliability is '80.95238095238095'. drift is 11.66507 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67572#0197842c-1b3f-458b-9e81-7d24c779dee0 | FLAKY         | 20/21    | upstream reliability is '95.65217391304348'. current run reliability is '95.23809523809523'. drift is 0.41408 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67754">build#67754</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi        | test_management            |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67754#01979e10-4960-4023-835f-b5e3d240420c | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| AuditLogTestKafkaApi        | test_management            |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67754#01979e12-28cc-4032-9a09-7d032b8622de | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67754#01979e10-4960-4023-835f-b5e3d240420c | FLAKY         | 17/21    | upstream reliability is '96.69811320754717'. current run reliability is '80.95238095238095'. drift is 15.74573 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67791">build#67791</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------------------|:-----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67791#0197a1d0-3d73-4906-9726-cd30a22bea16 | FLAKY         | 20/21    | upstream reliability is '98.9041095890411'. current run reliability is '95.23809523809523'. drift is 3.66601 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67791#0197a1d3-033b-46f6-9cf9-00720d7ab96d | FLAKY         | 13/15    |                                                                                                                                                                       |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67791#0197a1d3-033b-46f6-9cf9-00720d7ab96d | FLAKY         | 13/14    |                                                                                                                                                                       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67892">build#67892</a></summary>


| test_class                        | test_method                               | test_arguments                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:------------------------------------------|:------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest     | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67892#0197a7f5-e3f0-4bf4-9e93-8ffd8653211e | FLAKY         | 13/21    | upstream reliability is '70.57356608478803'. current run reliability is '61.904761904761905'. drift is 8.6688 and the allowed drift is set to 50. The test should PASS  |
| PartitionForceReconfigurationTest | test_basic_reconfiguration                | {"acks": -1, "controller_snapshots": true, "restart": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67892#0197a7f7-0397-4f4c-a091-17cfae98da4f | FLAKY         | 20/21    | upstream reliability is '99.36908517350159'. current run reliability is '95.23809523809523'. drift is 4.13099 and the allowed drift is set to 50. The test should PASS  |
| DisablingPartitionsTest           | test_disable                              |                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67892#0197a7f7-0399-4afe-94b1-0f4eef24ad32 | FLAKY         | 17/21    | upstream reliability is '92.74611398963731'. current run reliability is '80.95238095238095'. drift is 11.79373 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/security/acl.h:
```diff
@@ -291,6 +305,44 @@ class acl_principal final
       : _type(type)
       , _name(std::move(name)) {}
 
+    static acl_principal from_string(std::string_view principal) {
```
since it's moving, it seems like an excellent opportunity to get it out of the header and into a compiled .cc file.

dotnwat commented on src/v/pandaproxy/schema_registry/requests/acls.h:
```diff
@@ -0,0 +1,269 @@
+    std::string_view field_name;
+};
+
+struct acl {
```
can a lot of what's in sr/requests/acls.h be moved out of headers?

IoannisRP commented on src/v/security/acl.h:
```diff
@@ -35,6 +35,24 @@
+    ss::sstring msg;
+};
+
+template<typename E>
+std::enable_if_t<std::is_enum_v<E>, std::optional<E>>
+  from_string_view(std::string_view);
+
```
Nit:

this `enable_if` might be a bit too permissive. Could we use a concept to only allow the specific types we support further down? something like `is_acl_enum`?

Given that this is namespaced, it's probably not a big deal, but if the stars align, there would be lots of extra noise in the compiler error message.

pgellert replied:
Agree that it would be nicer, but I think it's not a big deal because I don't think the error message would be that bad. It would say that `from_string_view` is declared for some enum but its implementation is not found, which would prompt the person to define it, which I think is reasonable. A concept would lead to a nicer error message, but I'm not sure it's worth maintaining a list of enums here.

Fwiw, this is consistent with other places in the codebase: https://github.com/redpanda-data/redpanda/blob/f6bc66aeb509a62caeeb0d0ca71b19bdf942ed0b/src/v/pandaproxy/schema_registry/types.h#L41-L43

IoannisRP replied:
> It would say that from_string_view is declared for some enum but its implementation is not found, which would prompt the person to define it, which I think is reasonable

The problem is that this would be buried withing all the "Note: template function ... ignored" with all the existing types. Either way, not a big deal.

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -39,6 +50,59 @@ namespace pandaproxy::schema_registry {
+      perm, "ACL permission");
+}
+
+inline security::acl_host to_acl_host(const ss::sstring& host) {
+    if (host == "*") {
+        return security::acl_host::wildcard_host();
+    }
+    try {
+        return security::acl_host(host);
+    } catch (const std::invalid_argument& e) {
+        throw exception(
+          error_code::acl_invalid, fmt::format("Invalid ACL host: {}", host));
+    }
+}
+
+inline security::acl_principal to_acl_principal(const ss::sstring& principal) {
```
nit:

do the `inline` here have any purpose? we are in a nameless namespace in a `.cc` file 🤔

pgellert replied:
Move the code now and removed the `inline`

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -39,6 +50,59 @@ namespace pandaproxy::schema_registry {
+    try {
+        return security::acl_principal::from_string(principal);
+    } catch (const security::acl_conversion_error& e) {
+        throw exception(
+          error_code::acl_invalid, fmt::format("Invalid ACL principal: {}", e));
+    }
```
nit:

This error would read "Invalid ACL principal: Invalid principal name: {}"
And `acl_conversion_error` is already an `exception`. Not sure if this conversion is worth the extra cost...

pgellert replied:
The conversion to a `schema_registry::exception` is needed because it impacts the response behaviour (5xx vs 4xx status code). But good point about the error message, I've dropped the prefix now.

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +846,185 @@ status_ready(server::request_t rq, server::reply_t rp) {
+                               const std::string& param_name, auto converter) {
+        auto str_value = parse::query_param<std::optional<ss::sstring>>(
+          *rq.req, param_name);
+        return str_value ? std::make_optional(converter(*str_value))
+                         : std::nullopt;
```
nit:

In general, we seem to go with `has_value` and `value` instead of implicit conversions and de-referencing.

pgellert replied:
Moved to `parse_and_convert` now.

(I think fairly inconsistent here. We use a mix of both styles, depending on how verbose we want to be depending on the surrounding code.)

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +846,169 @@ status_ready(server::request_t rq, server::reply_t rp) {
+    auto permission = parse_and_convert("permission", to_acl_permission);
+    auto host = parse_and_convert("host", to_acl_host);
+
+    auto principal = principal_str.has_value()
+                       ? std::make_optional(to_acl_principal(*principal_str))
+                       : std::nullopt;
```
can't this use `parse_and_convert` as well?

pgellert replied:
Fixed

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +846,185 @@ status_ready(server::request_t rq, server::reply_t rp) {
+        security::resource_pattern_filter::resource_subsystem::schema_registry},
+      security::acl_entry_filter{principal, host, operation, permission}};
+
+    auto acls = acl_store.acls(filter);
+
+    auto sr_acls = std::vector<acl>{};
+    std::ranges::transform(
+      acls,
+      std::back_inserter(sr_acls),
+      [](const security::acl_binding& binding) { return acl(binding); });
```
nit:

you can use c++23 `std::ranges::to<std::vector>` to merge these 3 into a single command. something like

```suggestion
    auto sr_acls = std::ranges::to<std::vector>(
        acl_store.acls(filter) | std::views::transform(
            [](const security::acl_binding& binding) { return acl(binding); }
        )
    );
```

pgellert replied:
🔥 Thanks, updated

michael-redpanda replied:
oh that's hot

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +846,185 @@ status_ready(server::request_t rq, server::reply_t rp) {
+
+    check_feature_ready(rq);
+
+    auto err_vec = co_await security_frontend.create_acls(bindings, 5s);
```
Nit:

This seems to be the 3rd time we define the `5s` timeout in a `create_acls` call in pandaproxy (there are more re-definitions outside of pandaproxy).

No need to make any changes here. Just bringing it up as something to potentially fix in a later PR.

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +846,185 @@ status_ready(server::request_t rq, server::reply_t rp) {
+            to_acl_permission(acl.permission)});
+    }
+
+    check_feature_ready(rq);
```
Across all 3 new endpoints, wouldn't it make more sense to call this first thing before going through all the parsing? Or is there a reason to do this check after all the parsing is successful?

pgellert replied:
I chose to defer the feature check to prefer input validation errors over feature disabled errors. But I don't feel strongly either way.

michael-redpanda replied:
My $0.02 is that I like this pattern - check the input first, then check if we can do it

IoannisRP reviewed (commented):
Overall looks great! Mainly just nits...

pgellert commented:
force-push:
* [first](https://github.com/redpanda-data/redpanda/compare/5932989f3966ff2a4e33351b69c4233956daef26..73cdac9375dfa2c1403b7286b9dfcd61581472a2): rebase to dev to resolve merge conflicts on the feature table
* [second](https://github.com/redpanda-data/redpanda/compare/73cdac9375dfa2c1403b7286b9dfcd61581472a2..f6d5933abd48c7d68b233039e82e120e0b49e244): address PR feedback
* [third](https://github.com/redpanda-data/redpanda/compare/f6d5933abd48c7d68b233039e82e120e0b49e244..617e7f538f76d79f6a008bbff2d1aefc8d622835): noop, only updates commit messages

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/617e7f538f76d79f6a008bbff2d1aefc8d622835..2fb791b271d8ff01a24e9af588504fd0a510bef0): fix clang formatting

pgellert commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -5661,3 +5667,425 @@ def test_compatible_schemas(self, schema_type: str, mode: str):
+                                  operation="READ"),
+            self._create_test_acl(principal="User:bob",
+                                  resource="*",
+                                  resource_type="GLOBAL",
```
note to reviewers: I'll raise a follow up PR for the renaming of the GLOBAL resource type to REGISTRY to reduce the scope of this one

michael-redpanda commented on src/v/security/acl.h:
```diff
@@ -223,11 +237,11 @@ enum class principal_type : int8_t {
+        return "User";
     case principal_type::ephemeral_user:
-        return "ephemeral_user";
+        return "Ephemeral user";
```
nit: should this be `EphemeralUser` (same case and style as `RedpandaRole`)

pgellert replied:
I left it as `Ephemeral user` to leave it as it was formatted on kafka responses today (see the old code in src/v/kafka/server/handlers/details/security.h)

michael-redpanda commented on src/v/security/acl.cc:
```diff
@@ -380,8 +405,7 @@ std::ostream& operator<<(std::ostream& os, principal_type type) {
 operator<<(std::ostream& os, const acl_principal_base& principal) {
-    fmt::print(
-      os, "type {{{}}} name {{{}}}", principal.type(), principal.name_view());
+    fmt::print(os, "{}:{}", principal.type(), principal.name_view());
```
nice

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67754-->
<!--85e41ea8-7d01-4644-930f-fc131e800f33-->
<!--b2a91cc8-54a4-46df-8cd8-d58af55fe678-->
<!--title-->
## Retry command for [Build#67754](https://buildkite.com/redpanda/redpanda/builds/67754)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestKafkaApi.test_management
```
<!--/body-->

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +793,139 @@ status_ready(server::request_t rq, server::reply_t rp) {
+        security::resource_pattern_filter::resource_subsystem::schema_registry},
+      security::acl_entry_filter{principal, host, operation, permission}};
+
+    auto sr_acls = std::ranges::to<std::vector>(
+      acl_store.acls(filter)
+      | std::views::transform(
+        [](const security::acl_binding& binding) { return acl(binding); }));
```
this is so cool

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -782,4 +793,139 @@ status_ready(server::request_t rq, server::reply_t rp) {
+
+    auto raw_acls = co_await rjson_parse(
+      *rq.req, acl_handler<>{acl_handler<>::require_fields::yes});
+
```
should we just short-cicuit here with a 200 if no items are in the request?

pgellert replied:
I don't think it's necessary. `security_frontend.create_acls(...)` already short-circuits if there are no ACLs to create (same for delete), so the path to response is already minimal. I think we also want to respect the feature check even if the ACLs are empty, so we don't want to short-circuit before that.

michael-redpanda commented on src/v/pandaproxy/schema_registry/service.h:
```diff
@@ -27,6 +27,8 @@
 #include <seastar/core/smp.hh>
 #include <seastar/net/socket_defs.hh>
 
+#include <memory>
+
```
nit: probably unused

michael-redpanda replied:
or included earlier

pgellert replied:
It's for the `std::unique_ptr` I added. It's probably transitively included already, but in the spirit of IWYU I think it makes sense to add it.

michael-redpanda reviewed (dismissed):
all my comments are nits - you're free to make any changes/respond to them, but this overall lgtm

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/2fb791b271d8ff01a24e9af588504fd0a510bef0..6df8ad626fd1c97c83fdfae436135f4d3b38c172): fix failing audit logging test by using a context-aware fmt::formatter for acl_principal, instead of changing its formatting in-place. The earlier change unintentionally affected audit log output, which the test caught. To avoid modifying the audit format, I introduced two formatting modes based on context.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/6df8ad626fd1c97c83fdfae436135f4d3b38c172..9a65d4c32f8b585743ec79d11ec70ce197274c3b): rebase to dev to fix merge conflicts (after change in header include syntax)

pgellert committed:
security/acls: shorten SR resource names

To allow reusing the string formatting on the SR ACLs API.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/9a65d4c32f8b585743ec79d11ec70ce197274c3b..a462e593396b5424972b75acf272179486ed110c): fix unintended renames during refactor

michael-redpanda commented:
Final run passed - failed due to some infra related issues... force merging
